### PR TITLE
Let calico-bpf decide whether to fail when some counters error - XDP counters may not be present

### PR DESCRIPTION
**Pull request by mazdakn**
## Description
Failing in dumping or flushing a counter map should not prevent the operation for the rest of maps. So instead of returning an error, we just need to log and continue with the rest of maps.

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
TBD
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

### CODE REVIEW

File: felix/bpf/counters/counters.go [hunk 1]
```diff
@@ -164,7 +164,11 @@ func (c Counters) Read() ([][]uint64, error) {
...
 		val, err := c.read(c.maps[hook])
 		if err != nil {
-			return values, fmt.Errorf("Failed to read bpf counters. hook=%s err=%v", name, err)
+			if name == bpf.HookXDP {
+				logrus.Infof("Failed to read XDP bpf counters. err=%v", err)
+				continue
+			}
+			return values, fmt.Errorf("Failed to read bpf counters. hook=%s err=%w", name, err)
 		}
 		if len(values[hook]) < MaxCounterNumber {
```
Comment (tomastigera) [inline felix/bpf/counters/counters.go:173]: Is this ever possible given that on line 161 you allocate exactly this much space?

I think it is meant to test `len(val)`  and not to allocate at 161

**Review by tomastigera — CHANGES_REQUESTED** [review]
It feels like the `Read()` logic should be reserved to the user and not the lib/pkg and the pkg could just export the `read()` but with a few defined constants for the hooks. Let the user know that the XDP failed and the user can decide what to do about that.

I think that at line 182 `k := make([]byte, counterMapKeySize)` should not allocate but there should be a static pre-created key, that would also save the comment.

File: felix/cmd/calico-bpf/commands/counters.go [hunk 2]
```diff
@@ -99,11 +99,25 @@ func dumpInterface(cmd *cobra.Command, iface string) error {
...
-
 	bpfCounters := counters.NewCounters(iface)
-	values, err := bpfCounters.Read()
-	if err != nil {
-		return fmt.Errorf("Failed to read bpf counters. iface=%s err=%v", iface, err)
+
+	values := make([][]uint64, len(bpf.Hooks))
+	for i := range values {
+		values[i] = make([]uint64, counters.MaxCounterNumber)
+	}
```
Comment (tomastigera) [inline felix/cmd/calico-bpf/commands/counters.go:107]: why this allocation?

**Review by tomastigera — COMMENTED** [review]
LGTM, one nit :crossed_fingers: the CI passes :)

File: felix/cmd/calico-bpf/commands/counters.go (hunk 2 above)
Comment (mazdakn) [reply felix/cmd/calico-bpf/commands/counters.go:107]: `make` only creates the `[][]byte' but each entry `[]byte` is still `nil` thus accessing it panics. It's seems mandatory to also init each of `[]byte` in the 2D `[][]byte`.

Comment (tomastigera) [reply felix/cmd/calico-bpf/commands/counters.go:107]: But it is then assigned `values[index] = val` sure you would need to check if it is set at `newRow = append(newRow, fmt.Sprintf("%v", values[index][c.Counter]))` but isn't that likely what you want anyway when say the XDP counters are missing?

File: felix/cmd/calico-bpf/commands/counters.go [hunk 3]
```diff
@@ -99,11 +99,25 @@ func dumpInterface(cmd *cobra.Command, iface string) error {
...
+				log.Infof("Failed to read XDP bpf counters. err=%v", err)
+				continue
+			}
+			return fmt.Errorf("Failed to read bpf counters. iface=%v hook=%s err=%w", iface, hook, err)
+		}
+		if len(val) < counters.MaxCounterNumber {
+			return fmt.Errorf("Failed to read enough data from bpf counters. iface=%v hook=%s", iface, hook)
+		}
+		values[index] = make([]uint64, counters.MaxCounterNumber)
+		values[index] = val
```
Comment (tomastigera) [inline felix/cmd/calico-bpf/commands/counters.go:120]:
```suggestion
	values := make([][]uint64, len(bpf.Hooks))
	for index, hook := range bpf.Hooks {
		val, err := bpfCounters.Read(index)
		if err != nil {
			if hook == bpf.HookXDP {
				log.Infof("Failed to read XDP bpf counters. err=%v", err)
				continue
			}
			return fmt.Errorf("Failed to read bpf counters. iface=%v hook=%s err=%w", iface, hook, err)
		}
		if len(val) < counters.MaxCounterNumber {
			return fmt.Errorf("Failed to read enough data from bpf counters. iface=%v hook=%s", iface, hook)
		}
		values[index] = val
```

File: felix/cmd/calico-bpf/commands/counters.go [hunk 4]
```diff
@@ -114,8 +128,12 @@ func dumpInterface(cmd *cobra.Command, iface string) error {
 	for _, c := range counters.Descriptions() {
 		newRow := []string{c.Category, c.Caption}
 		// Now add value related to each hook, i.e. ingress, egress and XDP
-		for index := range bpf.Hooks {
-			newRow = append(newRow, fmt.Sprintf("%v", values[index][c.Counter]))
+		for index, hook := range bpf.Hooks {
+			if hook == bpf.HookXDP && !xdpOK {
```
Comment (tomastigera) [inline felix/cmd/calico-bpf/commands/counters.go:132]:
```suggestion
			if values[index] == nil {
```

**Review by tomastigera — COMMENTED** [review]
I think this is a little simpler :-)

```
		values[index] = make([]uint64, counters.MaxCounterNumber)
		values[index] = val
```

I wonder what static checks say about this assignment ;-)

**Review by tomastigera — APPROVED** [review]
pls squash

**Pull request merged**